### DISCUSSION TITLE
Chroma does all this

### DISCUSSION THREAD

**Discussion by ndrean** [dwyl/image-classifier#112] [General]
Check [this link](https://github.com/chroma-core/chroma#embeddings)

> If you search your photos for "famous bridge in San Francisco". By embedding this query and comparing it to the embeddings of your photos and their metadata - it should return photos of the Golden Gate Bridge.

**Comment by nelsonic** [comment]
@ndrean did you open this discussion as a way of comparing features of existing projects...? 🤷‍♂️ 

Yes. There will _always_ be multiple solutions to all problems. 👌 
McDonalds already does burgers. 🍔 ✅ 
Doesn't mean I'm not going to make my own gourmet ones from scratch to learn how to do it _better_. 😉 

We _only_ wanted a way of describing images we want to upload. 📸 ⬆️ ☁️ 📝 
The rest of the features are useless to us. 😒 
`Chroma` certainly looks good but we didn't _want_ to build ours in `Python` or `JavaScript`. 🙅

**Comment by ndrean** [comment]
> We only wanted a way of describing images we want to upload. 📸 ⬆️ ☁️ 📝
The rest of the features are useless to us. 😒

I understand that you mean that the semantic search is useless to you as a tool. I'm fine with this, but this repo demonstrated a way to run a semantic search in Elixir as embeddings and similarity search are basic AI/ML related tools.

> did you open this discussion as a way of comparing features of existing projects?

In a sense yes, but it goes beyond.
You may be curious about RAG [Retrieval Augmented Generation](https://aws.amazon.com/what-is/retrieval-augmented-generation/#:~:text=Retrieval%2DAugmented%20Generation%20(RAG),sources%20before%20generating%20a%20response.). Suppose you want to deliver a "natural language" response based on say your data, then you may need to store embeddings and run similarity search to inject this into the context of the LLM prompt. I wanted to point out that the `Chroma` project offers this enhanced vector badatase capabilities and is based on PG_VECTOR which I ruled out to use HNSWLib. It is Python based indeed, and my intention is more to draw attention to this particular solution, even if not in the Elixir world.

**Comment by ndrean** [comment]
Continuation of this RAG process that highlights vector search. You can find numerous papers on RAG.  A selection below.

- Firstly 2 pictures below illustrates how RAG is used to enhance LLM chatbots. 

 1)Extracted from the NVIDIA blog and I share the link. 

[<img width="1279" alt="Screenshot 2024-04-25 at 08 48 14" src="https://github.com/dwyl/image-classifier/assets/6793008/3914707b-5c3c-4a51-8a4d-d061d70a45ff">](https://blogs.nvidia.com/blog/what-is-retrieval-augmented-generation/)


  It says it reduces hallucination and claims to be easy (!).
  It also removes (?) the need of training your model. The later seems to be a hard task in the Elixir world, cf <https://elixirforum.com/t/rag-app-using-elixir-feasible/60439/16>).

 2) Another picture to illustrate where vector search (aka semantic search) is used in RAG process: you enhance the question with a context found by vector search from some controlled data, and let the LLM formulate the response nicely for you.

<img width="1125" alt="Screenshot 2024-04-25 at 12 06 54" src="https://github.com/dwyl/image-classifier/assets/6793008/e47ca25f-b43a-4c7b-8201-ab34fba78b7d">



- [a paper on RAG from Meta including a HuggingFace link](https://ai.meta.com/blog/retrieval-augmented-generation-streamlining-the-creation-of-intelligent-natural-language-processing-models/).


- Lastly, a [valuable resource](https://www.pinecone.io/learn/vector-database/) from the <https://pincecone.io> blog that explains the difference between a standalone Vector Index (such as `hnswlib`)  vs Vector Database using indexing (such as  `pg_vector` or `chromadb`). Similarity search is at the heart of these technologies. See [this other Pinecone blog on vector search](https://www.pinecone.io/learn/structured-data/). See [this link](https://www.aporia.com/learn/best-vector-dbs-for-retrieval-augmented-generation-rag/) for various managed vector db. It does not include `chromadb` which can be self hosted (in the same way as Postgres pg_vector).

I will try to experiment this in a Livebook. TODO
